File: leetCode/findMedianAlgo/findMedian.js
/*

  **MEDIAN OF TWO SORTED ARRAYS**

  Given two sorted arrays nums1 and nums2 of size m and n respectively, return the median of the two sorted arrays.

  The overall run time complexity should be O(log (m+n)).
*/

const nums1 = [1,2,3,5,7,8,9,10,13,15,16,17,19,65,77,79,80,99,101,104,108];
const nums2 = [2,66,88,96];

// My initial solution

// const findMedian = (nums1, nums2) => {
//   const merged = [];

//   nums1.map((num) => merged.push(num));
//   nums2.map((num) => merged.push(num));
//   merged.sort((a,b) => a - b);

//   const mid = Math.floor((merged.length - 1) / 2);

//   if (merged.length % 2 === 0) return (merged[mid] + merged[mid + 1] / 2);
//   else return merged[mid];

// }

// Solution with better joining method

const findMedian = (nums1, nums2) => {
  const merged = [];
  let i = 0, j = 0;

  while (i < nums1.length && j < nums2.length) {
    if (nums1[i] < nums2[j]) {
      merged.push(nums1[i++])
    } else {
      merged.push(nums2[j++])
    }
  }

  while (i < nums1.length) merged.push(nums1[i++])
  while (j < nums2.length) merged.push(nums2[j++])

  const mid = Math.floor((merged.length - 1) / 2);

  if (merged.length % 2 === 0) return (merged[mid] + merged[mid + 1] / 2);
  else return merged[mid];

}

console.log(findMedian(nums1, nums2));

File: leetCode/findMedianAlgo/findMedian.py
from typing import List
import math

nums1 = [1,2,3,5,7,8,9,10,13,15,16,17,19,65,77,79,80,99,101,104,108];
nums2 = [2,66,88,96,109];

def findMedianSortedArrays(nums1: List[int], nums2: List[int]) -> float:
  merged = []
  i = 0
  j = 0

  while i < len(nums1) and j < len(nums2):
    if nums1[i] < nums2[j]:
      merged.append(nums1[i])
      i += 1
    else:
      merged.append(nums2[j])
      j += 1

  while i < len(nums1):
    merged.append(nums1[i])
    i += 1

  while j < len(nums2):
    merged.append(nums2[j])
    j += 1

  mid = math.floor(len(merged) / 2)

  if (len(merged) % 2 == 0):
    return (merged[mid - 1] + merged[mid]) / 2
  else:
    return merged[mid]
  
print(findMedianSortedArrays(nums1, nums2))